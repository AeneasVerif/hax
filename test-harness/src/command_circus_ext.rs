use lazy_static::{__Deref, lazy_static};
use std::{ffi::OsStr, process::Command};

pub trait CommandCircusExt {
    fn circus<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Self;
}

impl CommandCircusExt for Command {
    fn circus<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Command {
        use assert_cmd::cargo::cargo_bin;
        use std::path::PathBuf;
        struct Paths {
            driver: PathBuf,
            engine: PathBuf,
            cargo_circus: PathBuf,
        }
        const CARGO_CIRCUS: &str = "cargo-circus";
        lazy_static! {
            static ref PATHS: Option<Paths> = {
                if let "yes" | "y" | "true" | "1" = std::env::var("CARGO_TESTS_ASSUME_BUILT").unwrap_or("".into()).to_lowercase().as_str() {
                    return None;
                }
                let root = std::env::current_dir().unwrap();
                let root = root.parent().unwrap();
                let engine_dir = root.join("engine");
                // Make sure binaries are built
                assert!(Command::new("cargo")
                        .args(&["build", "--workspace", "--bins"])
                        .status()
                        .unwrap()
                        .success());
                assert!(Command::new("dune")
                        .args(&["build"])
                        .env("CIRCUS_JSON_SCHEMA_EXPORTER_BINARY", cargo_bin("circus-export-json-schemas"))
                        .current_dir(engine_dir.clone())
                        .status()
                        .unwrap()
                        .success());
                Some(Paths {
                    driver: cargo_bin("driver-circus-frontend-exporter"),
                    cargo_circus: cargo_bin(CARGO_CIRCUS),
                    engine: engine_dir.join("_build/install/default/bin/circus-engine"),
                })
            };
        }
        let mut cmd = match PATHS.deref() {
            Some(paths) => {
                let mut cmd = Command::new(paths.cargo_circus.clone());
                cmd.env("CIRCUS_RUSTC_DRIVER_BINARY", paths.driver.clone());
                cmd.env("CIRCUS_ENGINE_BINARY", paths.engine.clone());
                cmd
            }
            None => Command::new(CARGO_CIRCUS),
        };
        cmd.args(args);
        // As documented in
        // https://doc.rust-lang.org/cargo/reference/environment-variables.html#dynamic-library-paths,
        // [cargo run] (and thus also [cargo test]) sets dynamic
        // library paths, which causes some issues with dependencies
        // when compiling without rustup
        for env in ["DYLD_FALLBACK_LIBRARY_PATH", "LD_LIBRARY_PATH"] {
            cmd.env_remove(env);
        }
        cmd
    }
}
