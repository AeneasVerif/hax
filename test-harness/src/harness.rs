mod command_circus_ext;
use command_circus_ext::*;
use serde_json::Value;
use std::process::{Command, Stdio};

#[derive(Clone, Debug)]
pub enum TestKind {
    Translate { backend: String },
    Lint { linter: String },
}

impl TestKind {
    fn as_subcommands(&self) -> [String; 2] {
        match self {
            TestKind::Lint { linter } => ["lint".to_string(), linter.clone()],
            TestKind::Translate { backend } => ["into".to_string(), backend.clone()],
        }
    }
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct TestSpec {
    /// a slow test can be made optional
    pub optional: bool,
    /// a broken test a test that should succeed (or fail) but does
    /// not dues to a bug to be fixed (see field [issue_id] below)
    pub broken: bool,
    /// Github issue ID
    pub issue_id: Option<u64>,
    /// Is that a positive or a negative test?
    pub positive: bool,
}
impl From<Value> for TestSpec {
    /// Parse a JSON value into a TestSpec
    fn from(o: Value) -> Self {
        let as_bool = |key: &str, default: bool| {
            let b = &o[key];
            if b.is_null() {
                default
            } else {
                b.as_bool().unwrap()
            }
        };
        Self {
            optional: as_bool("optional", false),
            positive: as_bool("positive", true),
            broken: as_bool("broken", false),
            issue_id: o["issue_id"].as_u64(),
        }
    }
}

/// The information for a test is given by `cargo metadata`
#[derive(Clone, Debug)]
pub struct TestInfo {
    pub name: String,
    pub manifest: std::path::PathBuf,
    pub description: Option<String>,
}

#[derive(Clone, Debug)]
pub struct Test {
    pub kind: TestKind,
    pub info: TestInfo,
    pub spec: TestSpec,
}

impl std::fmt::Display for Test {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{} - {:?}", self.info.name, self.kind)?;
        if let Some(issue_id) = self.spec.issue_id {
            write!(f, " #{}", issue_id)?;
        };
        Ok(())
    }
}

impl Test {
    fn into_runner(self, workspace: String) -> Result<(), libtest_mimic::Failed> {
        // 1. cook a command
        let mut cmd = Command::circus(&["-C"]);
        cmd.arg("--manifest-path").arg(self.info.manifest.clone());
        cmd.arg(";");
        cmd.stdout(Stdio::piped()).stderr(Stdio::piped());
        cmd.args(self.kind.as_subcommands());

        // 2. execute it (twice, idea of @franziskuskiefer, so that
        // the messages related to building dependencies are not
        // included in the second one)
        let _ = cmd.output().unwrap();
        let out = cmd.output().unwrap();

        let command_successful = out.status.success();
        let cleanup = |s: String| {
            s.replace(r"\", "/")
                .replace(&workspace, "WORKSPACE_ROOT")
                .replace("Blocking waiting for file lock on build directory", "")
                .trim()
                .to_string()
        };
        let serr = cleanup(String::from_utf8_lossy(&out.stderr).to_string());
        let sout = cleanup(String::from_utf8_lossy(&out.stdout).to_string());

        // 3. make sure the test is successful
        let err = |s: &str| {
            Err(format!(
                "Command {s}.\nThe command was: {:?}{}",
                cmd,
                if command_successful {
                    "".to_string()
                } else {
                    format!("\nSTDOUT:\n{}\nSTDERR:\n{}", sout, serr)
                }
            ))
        };
        match (command_successful, (self.spec.positive, self.spec.broken)) {
            (false, (false, false) | (true, true)) => {
                #[derive(serde::Serialize)]
                struct Output {
                    serr: String,
                    sout: String,
                    exit: i32,
                }
                let exit = out.status.code().unwrap_or(std::i32::MAX);
                let out = &Output { serr, sout, exit };
                let name = format!(
                    "{}+{}",
                    self.info.name,
                    self.kind.as_subcommands().join("-")
                );
                insta::with_settings!({
                    info => &self.spec,
                }, { Ok(insta::assert_toml_snapshot!(name, out)) })
            }
            (false, (false, true)) => err("failed, but this is a negative test marked broken")?,
            (false, (true, false)) => err("failed")?,
            (true, (true, false) | (false, true)) => Ok(()),
            (true, (true, true)) => err("succeeded, but this is a positive test marked broken")?,
            (true, (false, false)) => err("succeeded, but this is a negative test")?,
        }
    }

    fn into_trial(&self, workspace: &String) -> libtest_mimic::Trial {
        libtest_mimic::Trial::test(format!("{}", &self), {
            let test = self.clone();
            let workspace = workspace.clone();
            move || test.clone().into_runner(workspace)
        })
        .with_kind(if self.spec.positive {
            "positive"
        } else {
            "negative"
        })
        .with_ignored_flag(self.spec.optional)
    }
}

/// Given [metadata] the table declared in a test's [Cargo.toml]
/// [workspace.circus-tests], this function returns a list of tests
fn parse_circus_tests_metadata(info: TestInfo, metadata: &Value) -> Vec<Test> {
    metadata
        .as_object()
        .expect("Expected value at key [circus-tests] to be a dictionary")
        .into_iter()
        .flat_map(|(a, o)| {
            o.as_object()
                .expect(format!("Expected value at key [{}] be a dictionary", a).as_str())
                .into_iter()
                .flat_map(|(key, o)| key.split("+").map(|k| (k.trim().to_string(), o.clone())))
                .map(|(b, o)| (a.clone(), b, o))
        })
        .map(|(a, b, o)| Test {
            spec: o.into(),
            info: info.clone(),
            kind: match a.as_str() {
                "into" => TestKind::Translate { backend: b },
                "lint" => TestKind::Lint { linter: b },
                _ => panic!("unexpected metadata [circus-tests.{}.{}]", a, b),
            },
        })
        .collect()
}

fn main() {
    let metadata = cargo_metadata::MetadataCommand::new()
        .manifest_path("../tests/Cargo.toml")
        .exec()
        .unwrap();
    let workspace_root: String = metadata.workspace_root.into();

    libtest_mimic::run(
        &libtest_mimic::Arguments::from_args(),
        metadata
            .packages
            .into_iter()
            .flat_map(|o| {
                parse_circus_tests_metadata(
                    TestInfo {
                        name: o.name,
                        description: o.description,
                        manifest: o.manifest_path.into(),
                    },
                    &o.metadata["circus-tests"],
                )
            })
            .map(|test| test.into_trial(&workspace_root))
            .collect(),
    )
    .exit();
}
