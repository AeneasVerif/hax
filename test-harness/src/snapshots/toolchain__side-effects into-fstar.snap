---
source: test-harness/src/harness.rs
expression: snapshot
info:
  kind:
    Translate:
      backend: fstar
  info:
    name: side-effects
    manifest: side-effects/Cargo.toml
    description: ~
  spec:
    optional: false
    broken: false
    issue_id: ~
    positive: true
    snapshot:
      stderr: true
      stdout: true
    include_flag: ~
    backend_options: ~
---
exit = 0
stderr = 'Finished `dev` profile [unoptimized + debuginfo] target(s) in XXs'

[stdout]
diagnostics = []

[stdout.files]
"Side_effects.Issue_1083_.fst" = '''
module Side_effects.Issue_1083_
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

class t_MyFrom (v_Self: Type0) (v_T: Type0) = {
  f_my_from_pre:v_T -> Type0;
  f_my_from_post:v_T -> v_Self -> Type0;
  f_my_from:x0: v_T -> Prims.Pure v_Self (f_my_from_pre x0) (fun result -> f_my_from_post x0 result)
}

[@@ FStar.Tactics.Typeclasses.tcinstance]
let impl: t_MyFrom u16 u8 =
  {
    f_my_from_pre = (fun (x: u8) -> true);
    f_my_from_post = (fun (x: u8) (out: u16) -> true);
    f_my_from = fun (x: u8) -> cast (x <: u8) <: u16
  }

let f (x: u8) : Core.Result.t_Result u16 u16 =
  match Core.Result.Result_Err (mk_u8 1) <: Core.Result.t_Result Prims.unit u8 with
  | Core.Result.Result_Ok _ ->
    Core.Result.Result_Ok (f_my_from #u16 #u8 #FStar.Tactics.Typeclasses.solve x)
    <:
    Core.Result.t_Result u16 u16
  | Core.Result.Result_Err err ->
    Core.Result.Result_Err (Core.Convert.f_from #u16 #u8 #FStar.Tactics.Typeclasses.solve err)
    <:
    Core.Result.t_Result u16 u16
'''
"Side_effects.Issue_1089_.fst" = '''
module Side_effects.Issue_1089_
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

let test (x y: Core.Option.t_Option i32) : Core.Option.t_Option i32 =
  match
    Core.Option.impl__map #i32
      #(Core.Option.t_Option i32)
      x
      (fun i ->
          let i:i32 = i in
          match y <: Core.Option.t_Option i32 with
          | Core.Option.Option_Some hoist1 ->
            Core.Option.Option_Some (i +! hoist1 <: i32) <: Core.Option.t_Option i32
          | Core.Option.Option_None  -> Core.Option.Option_None <: Core.Option.t_Option i32)
    <:
    Core.Option.t_Option (Core.Option.t_Option i32)
  with
  | Core.Option.Option_Some some -> some
  | Core.Option.Option_None  -> Core.Option.Option_None <: Core.Option.t_Option i32
'''
"Side_effects.Nested_return.fst" = '''
module Side_effects.Nested_return
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

let other_fun (rng: i8) : (i8 & Core.Result.t_Result Prims.unit Prims.unit) =
  let hax_temp_output:Core.Result.t_Result Prims.unit Prims.unit =
    Core.Result.Result_Ok (() <: Prims.unit) <: Core.Result.t_Result Prims.unit Prims.unit
  in
  rng, hax_temp_output <: (i8 & Core.Result.t_Result Prims.unit Prims.unit)

let v_fun (rng: i8) : (i8 & Core.Result.t_Result Prims.unit Prims.unit) =
  let tmp0, out:(i8 & Core.Result.t_Result Prims.unit Prims.unit) = other_fun rng in
  let rng:i8 = tmp0 in
  match out <: Core.Result.t_Result Prims.unit Prims.unit with
  | Core.Result.Result_Ok hoist6 ->
    rng, (Core.Result.Result_Ok hoist6 <: Core.Result.t_Result Prims.unit Prims.unit)
    <:
    (i8 & Core.Result.t_Result Prims.unit Prims.unit)
  | Core.Result.Result_Err err ->
    rng, (Core.Result.Result_Err err <: Core.Result.t_Result Prims.unit Prims.unit)
    <:
    (i8 & Core.Result.t_Result Prims.unit Prims.unit)
'''
"Side_effects.fst" = '''
module Side_effects
#set-options "--fuel 0 --ifuel 1 --z3rlimit 15"
open Core
open FStar.Mul

type t_A = | A : t_A

type t_B = | B : t_B

type t_Bar = {
  f_a:bool;
  f_b:(t_Array (bool & bool) (mk_usize 6) & bool)
}

type t_Foo = {
  f_x:bool;
  f_y:(bool & Alloc.Vec.t_Vec t_Bar Alloc.Alloc.t_Global);
  f_z:t_Array t_Bar (mk_usize 6);
  f_bar:t_Bar
}

/// Helper function
let add3 (x y z: u32) : u32 =
  Core.Num.impl__u32__wrapping_add (Core.Num.impl__u32__wrapping_add x y <: u32) z

/// Test assignation on non-trivial places
let assign_non_trivial_lhs (foo: t_Foo) : t_Foo =
  let foo:t_Foo = { foo with f_x = true } <: t_Foo in
  let foo:t_Foo = { foo with f_bar = { foo.f_bar with f_a = true } <: t_Bar } <: t_Foo in
  let foo:t_Foo =
    {
      foo with
      f_bar
      =
      {
        foo.f_bar with
        f_b
        =
        {
          foo.f_bar.f_b with
          _1
          =
          Rust_primitives.Hax.Monomorphized_update_at.update_at_usize foo.f_bar.f_b._1
            (mk_usize 3)
            ({ (foo.f_bar.f_b._1.[ mk_usize 3 ] <: (bool & bool)) with _2 = true } <: (bool & bool))
        }
        <:
        (t_Array (bool & bool) (mk_usize 6) & bool)
      }
      <:
      t_Bar
    }
    <:
    t_Foo
  in
  let foo:t_Foo =
    {
      foo with
      f_z
      =
      Rust_primitives.Hax.Monomorphized_update_at.update_at_usize foo.f_z
        (mk_usize 3)
        ({ (foo.f_z.[ mk_usize 3 ] <: t_Bar) with f_a = true } <: t_Bar)
    }
    <:
    t_Foo
  in
  let foo:t_Foo =
    {
      foo with
      f_y
      =
      {
        foo.f_y with
        _2
        =
        Rust_primitives.Hax.Monomorphized_update_at.update_at_usize foo.f_y._2
          (mk_usize 3)
          ({
              (foo.f_y._2.[ mk_usize 3 ] <: t_Bar) with
              f_b
              =
              {
                (foo.f_y._2.[ mk_usize 3 ] <: t_Bar).f_b with
                _1
                =
                Rust_primitives.Hax.Monomorphized_update_at.update_at_usize (foo.f_y._2.[ mk_usize 3
                    ]
                    <:
                    t_Bar)
                    .f_b
                    ._1
                  (mk_usize 5)
                  ({
                      ((foo.f_y._2.[ mk_usize 3 ] <: t_Bar).f_b._1.[ mk_usize 5 ] <: (bool & bool)) with
                      _1 = true
                    }
                    <:
                    (bool & bool))
                <:
                t_Array (bool & bool) (mk_usize 6)
              }
              <:
              (t_Array (bool & bool) (mk_usize 6) & bool)
            }
            <:
            t_Bar)
      }
      <:
      (bool & Alloc.Vec.t_Vec t_Bar Alloc.Alloc.t_Global)
    }
    <:
    t_Foo
  in
  foo

/// Question mark without error coercion
let direct_result_question_mark (y: Core.Result.t_Result Prims.unit u32)
    : Core.Result.t_Result i8 u32 =
  match y <: Core.Result.t_Result Prims.unit u32 with
  | Core.Result.Result_Ok _ -> Core.Result.Result_Ok (mk_i8 0) <: Core.Result.t_Result i8 u32
  | Core.Result.Result_Err err -> Core.Result.Result_Err err <: Core.Result.t_Result i8 u32

/// Question mark with an error coercion
let direct_result_question_mark_coercion (y: Core.Result.t_Result i8 u16)
    : Core.Result.t_Result i8 u32 =
  match y <: Core.Result.t_Result i8 u16 with
  | Core.Result.Result_Ok hoist10 -> Core.Result.Result_Ok hoist10 <: Core.Result.t_Result i8 u32
  | Core.Result.Result_Err err ->
    Core.Result.Result_Err (Core.Convert.f_from #u32 #u16 #FStar.Tactics.Typeclasses.solve err)
    <:
    Core.Result.t_Result i8 u32

/// Exercise early returns with control flow and loops
let early_returns (x: u32) : u32 =
  if x >. mk_u32 3
  then mk_u32 0
  else
    if x >. mk_u32 30
    then
      match true <: bool with
      | true -> mk_u32 34
      | _ ->
        let x, hoist14:(u32 & u32) = x, mk_u32 3 <: (u32 & u32) in
        Core.Num.impl__u32__wrapping_add (Core.Num.impl__u32__wrapping_add (mk_u32 123) hoist14
            <:
            u32)
          x
    else
      let x:u32 = x +! mk_u32 9 in
      let x, hoist14:(u32 & u32) = x, x +! mk_u32 1 <: (u32 & u32) in
      Core.Num.impl__u32__wrapping_add (Core.Num.impl__u32__wrapping_add (mk_u32 123) hoist14 <: u32
        )
        x

/// Exercise local mutation with control flow and loops
let local_mutation (x: u32) : u32 =
  let y:u32 = mk_u32 0 in
  let x:u32 = Core.Num.impl__u32__wrapping_add x (mk_u32 1) in
  if x >. mk_u32 3
  then
    let x:u32 = Core.Num.impl__u32__wrapping_sub x (mk_u32 3) in
    let y:u32 = x /! mk_u32 2 in
    let y:u32 = Core.Num.impl__u32__wrapping_add y (mk_u32 2) in
    let y:u32 =
      Core.Iter.Traits.Iterator.f_fold (Core.Iter.Traits.Collect.f_into_iter #(Core.Ops.Range.t_Range
              u32)
            #FStar.Tactics.Typeclasses.solve
            ({ Core.Ops.Range.f_start = mk_u32 0; Core.Ops.Range.f_end = mk_u32 10 }
              <:
              Core.Ops.Range.t_Range u32)
          <:
          Core.Ops.Range.t_Range u32)
        y
        (fun y i ->
            let y:u32 = y in
            let i:u32 = i in
            Core.Num.impl__u32__wrapping_add x i <: u32)
    in
    Core.Num.impl__u32__wrapping_add x y
  else
    let (x, y), hoist24:((u32 & u32) & u32) =
      match x <: u32 with
      | Rust_primitives.Integers.MkInt 12 ->
        let y:u32 = Core.Num.impl__u32__wrapping_add x y in
        (x, y <: (u32 & u32)), mk_u32 3 <: ((u32 & u32) & u32)
      | Rust_primitives.Integers.MkInt 13 ->
        let x:u32 = Core.Num.impl__u32__wrapping_add x (mk_u32 1) in
        (x, y <: (u32 & u32)), add3 x (Core.Num.impl__u32__wrapping_add (mk_u32 123) x <: u32) x
        <:
        ((u32 & u32) & u32)
      | _ -> (x, y <: (u32 & u32)), mk_u32 0 <: ((u32 & u32) & u32)
    in
    let x:u32 = hoist24 in
    Core.Num.impl__u32__wrapping_add x y

/// Combine `?` and early return
let monad_lifting (x: u8) : Core.Result.t_Result t_A t_B =
  if x >. mk_u8 123
  then
    match Core.Result.Result_Err (B <: t_B) <: Core.Result.t_Result t_A t_B with
    | Core.Result.Result_Ok hoist25 -> Core.Result.Result_Ok hoist25 <: Core.Result.t_Result t_A t_B
    | Core.Result.Result_Err err -> Core.Result.Result_Err err <: Core.Result.t_Result t_A t_B
  else Core.Result.Result_Ok (A <: t_A) <: Core.Result.t_Result t_A t_B

/// Test question mark on `Option`s with some control flow
let options (x y: Core.Option.t_Option u8) (z: Core.Option.t_Option u64) : Core.Option.t_Option u8 =
  match x <: Core.Option.t_Option u8 with
  | Core.Option.Option_Some hoist31 ->
    if hoist31 >. mk_u8 10
    then
      match x <: Core.Option.t_Option u8 with
      | Core.Option.Option_Some hoist33 ->
        (match
            Core.Option.Option_Some (Core.Num.impl__u8__wrapping_add hoist33 (mk_u8 3))
            <:
            Core.Option.t_Option u8
          with
          | Core.Option.Option_Some hoist39 ->
            (match hoist39 <: u8 with
              | Rust_primitives.Integers.MkInt 3 ->
                (match Core.Option.Option_None <: Core.Option.t_Option u8 with
                  | Core.Option.Option_Some some ->
                    let v:u8 = some in
                    (match x <: Core.Option.t_Option u8 with
                      | Core.Option.Option_Some hoist40 ->
                        (match y <: Core.Option.t_Option u8 with
                          | Core.Option.Option_Some hoist41 ->
                            Core.Option.Option_Some
                            (Core.Num.impl__u8__wrapping_add (Core.Num.impl__u8__wrapping_add v
                                    hoist40
                                  <:
                                  u8)
                                hoist41)
                            <:
                            Core.Option.t_Option u8
                          | Core.Option.Option_None  ->
                            Core.Option.Option_None <: Core.Option.t_Option u8)
                      | Core.Option.Option_None  ->
                        Core.Option.Option_None <: Core.Option.t_Option u8)
                  | Core.Option.Option_None  -> Core.Option.Option_None <: Core.Option.t_Option u8)
              | Rust_primitives.Integers.MkInt 4 ->
                (match z <: Core.Option.t_Option u64 with
                  | Core.Option.Option_Some hoist28 ->
                    let v:u8 =
                      mk_u8 4 +! (if hoist28 >. mk_u64 4 <: bool then mk_u8 0 else mk_u8 3)
                    in
                    (match x <: Core.Option.t_Option u8 with
                      | Core.Option.Option_Some hoist40 ->
                        (match y <: Core.Option.t_Option u8 with
                          | Core.Option.Option_Some hoist41 ->
                            Core.Option.Option_Some
                            (Core.Num.impl__u8__wrapping_add (Core.Num.impl__u8__wrapping_add v
                                    hoist40
                                  <:
                                  u8)
                                hoist41)
                            <:
                            Core.Option.t_Option u8
                          | Core.Option.Option_None  ->
                            Core.Option.Option_None <: Core.Option.t_Option u8)
                      | Core.Option.Option_None  ->
                        Core.Option.Option_None <: Core.Option.t_Option u8)
                  | Core.Option.Option_None  -> Core.Option.Option_None <: Core.Option.t_Option u8)
              | _ ->
                let v:u8 = mk_u8 12 in
                match x <: Core.Option.t_Option u8 with
                | Core.Option.Option_Some hoist40 ->
                  (match y <: Core.Option.t_Option u8 with
                    | Core.Option.Option_Some hoist41 ->
                      Core.Option.Option_Some
                      (Core.Num.impl__u8__wrapping_add (Core.Num.impl__u8__wrapping_add v hoist40
                            <:
                            u8)
                          hoist41)
                      <:
                      Core.Option.t_Option u8
                    | Core.Option.Option_None  -> Core.Option.Option_None <: Core.Option.t_Option u8
                  )
                | Core.Option.Option_None  -> Core.Option.Option_None <: Core.Option.t_Option u8)
          | Core.Option.Option_None  -> Core.Option.Option_None <: Core.Option.t_Option u8)
      | Core.Option.Option_None  -> Core.Option.Option_None <: Core.Option.t_Option u8
    else
      (match x <: Core.Option.t_Option u8 with
        | Core.Option.Option_Some hoist36 ->
          (match y <: Core.Option.t_Option u8 with
            | Core.Option.Option_Some hoist35 ->
              (match
                  Core.Option.Option_Some (Core.Num.impl__u8__wrapping_add hoist36 hoist35)
                  <:
                  Core.Option.t_Option u8
                with
                | Core.Option.Option_Some hoist39 ->
                  (match hoist39 <: u8 with
                    | Rust_primitives.Integers.MkInt 3 ->
                      (match Core.Option.Option_None <: Core.Option.t_Option u8 with
                        | Core.Option.Option_Some some ->
                          let v:u8 = some in
                          (match x <: Core.Option.t_Option u8 with
                            | Core.Option.Option_Some hoist40 ->
                              (match y <: Core.Option.t_Option u8 with
                                | Core.Option.Option_Some hoist41 ->
                                  Core.Option.Option_Some
                                  (Core.Num.impl__u8__wrapping_add (Core.Num.impl__u8__wrapping_add v
                                          hoist40
                                        <:
                                        u8)
                                      hoist41)
                                  <:
                                  Core.Option.t_Option u8
                                | Core.Option.Option_None  ->
                                  Core.Option.Option_None <: Core.Option.t_Option u8)
                            | Core.Option.Option_None  ->
                              Core.Option.Option_None <: Core.Option.t_Option u8)
                        | Core.Option.Option_None  ->
                          Core.Option.Option_None <: Core.Option.t_Option u8)
                    | Rust_primitives.Integers.MkInt 4 ->
                      (match z <: Core.Option.t_Option u64 with
                        | Core.Option.Option_Some hoist28 ->
                          let v:u8 =
                            mk_u8 4 +! (if hoist28 >. mk_u64 4 <: bool then mk_u8 0 else mk_u8 3)
                          in
                          (match x <: Core.Option.t_Option u8 with
                            | Core.Option.Option_Some hoist40 ->
                              (match y <: Core.Option.t_Option u8 with
                                | Core.Option.Option_Some hoist41 ->
                                  Core.Option.Option_Some
                                  (Core.Num.impl__u8__wrapping_add (Core.Num.impl__u8__wrapping_add v
                                          hoist40
                                        <:
                                        u8)
                                      hoist41)
                                  <:
                                  Core.Option.t_Option u8
                                | Core.Option.Option_None  ->
                                  Core.Option.Option_None <: Core.Option.t_Option u8)
                            | Core.Option.Option_None  ->
                              Core.Option.Option_None <: Core.Option.t_Option u8)
                        | Core.Option.Option_None  ->
                          Core.Option.Option_None <: Core.Option.t_Option u8)
                    | _ ->
                      let v:u8 = mk_u8 12 in
                      match x <: Core.Option.t_Option u8 with
                      | Core.Option.Option_Some hoist40 ->
                        (match y <: Core.Option.t_Option u8 with
                          | Core.Option.Option_Some hoist41 ->
                            Core.Option.Option_Some
                            (Core.Num.impl__u8__wrapping_add (Core.Num.impl__u8__wrapping_add v
                                    hoist40
                                  <:
                                  u8)
                                hoist41)
                            <:
                            Core.Option.t_Option u8
                          | Core.Option.Option_None  ->
                            Core.Option.Option_None <: Core.Option.t_Option u8)
                      | Core.Option.Option_None  ->
                        Core.Option.Option_None <: Core.Option.t_Option u8)
                | Core.Option.Option_None  -> Core.Option.Option_None <: Core.Option.t_Option u8)
            | Core.Option.Option_None  -> Core.Option.Option_None <: Core.Option.t_Option u8)
        | Core.Option.Option_None  -> Core.Option.Option_None <: Core.Option.t_Option u8)
  | Core.Option.Option_None  -> Core.Option.Option_None <: Core.Option.t_Option u8

/// Test question mark on `Result`s with local mutation
let question_mark (x: u32) : Core.Result.t_Result u32 u32 =
  if x >. mk_u32 40
  then
    let y:u32 = mk_u32 0 in
    let x:u32 = Core.Num.impl__u32__wrapping_add x (mk_u32 3) in
    let y:u32 = Core.Num.impl__u32__wrapping_add x y in
    let x:u32 = Core.Num.impl__u32__wrapping_add x y in
    if x >. mk_u32 90
    then
      match Core.Result.Result_Err (mk_u8 12) <: Core.Result.t_Result Prims.unit u8 with
      | Core.Result.Result_Ok ok ->
        Core.Result.Result_Ok (Core.Num.impl__u32__wrapping_add (mk_u32 3) x)
        <:
        Core.Result.t_Result u32 u32
      | Core.Result.Result_Err err ->
        Core.Result.Result_Err (Core.Convert.f_from #u32 #u8 #FStar.Tactics.Typeclasses.solve err)
        <:
        Core.Result.t_Result u32 u32
    else
      Core.Result.Result_Ok (Core.Num.impl__u32__wrapping_add (mk_u32 3) x)
      <:
      Core.Result.t_Result u32 u32
  else
    Core.Result.Result_Ok (Core.Num.impl__u32__wrapping_add (mk_u32 3) x)
    <:
    Core.Result.t_Result u32 u32

let simplifiable_question_mark (c: bool) (x: Core.Option.t_Option i32) : Core.Option.t_Option i32 =
  if c
  then
    match x <: Core.Option.t_Option i32 with
    | Core.Option.Option_Some hoist45 ->
      let a:i32 = hoist45 +! mk_i32 10 in
      let b:i32 = mk_i32 20 in
      Core.Option.Option_Some (a +! b) <: Core.Option.t_Option i32
    | Core.Option.Option_None  -> Core.Option.Option_None <: Core.Option.t_Option i32
  else
    let a:i32 = mk_i32 0 in
    let b:i32 = mk_i32 20 in
    Core.Option.Option_Some (a +! b) <: Core.Option.t_Option i32

let simplifiable_return (c1 c2 c3: bool) : i32 =
  let x:i32 = mk_i32 0 in
  if c1
  then
    if c2
    then
      let x:i32 = x +! mk_i32 10 in
      if c3 then mk_i32 1 else x +! mk_i32 1
    else x +! mk_i32 1
  else x
'''
